### ISSUE TITLE
python3: weak binary security

### ISSUE THREAD

**Issue by nrvnrvn** [wolfi-dev/os#556]
Hello! I have been evaluating chainguard images lately and noticed that the python binary does not feature Stack canaries, Full RELRO and FORTIFY_SOURCE.

### Steps to reproduce

```Dockerfile
$ cat <<EOF | docker build -
FROM cgr.dev/chainguard/python:latest-dev as cgr

FROM alpine:3.17 as alpine

RUN apk add python3

FROM ubuntu:jammy

COPY --from=cgr /usr/bin/python3.11 /check/python3.11-chainguard

COPY --from=alpine /usr/bin/python3.10 /check/python3.10-alpine

RUN apt-get update &&\
  apt-get install -y file checksec python3

RUN checksec --file=/check/python3.11-chainguard &&\
  file /check/python3.11-chainguard &&\
  ldd /check/python3.11-chainguard

RUN checksec --file=/check/python3.10-alpine &&\
  file /check/python3.10-alpine &&\
  ldd /check/python3.10-alpine

RUN checksec --file=/usr/bin/python3.10 &&\
  file /usr/bin/python3.10 &&\
  ldd /usr/bin/python3.10
EOF
```

#### for `cgr.dev/chainguard/python`
```
Step 8/10 : RUN checksec --file=/check/python3.11-chainguard &&  file /check/python3.11-chainguard &&  ldd /check/python3.11-chainguard
 ---> Running in 3bc694638b95
RELRO           STACK CANARY      NX            PIE             RPATH      RUNPATH	Symbols		FORTIFY	Fortified	Fortifiable	FILE
Partial RELRO   No canary found   NX enabled    PIE enabled     No RPATH   No RUNPATH   No Symbols	  No	0		0		/check/python3.11-chainguard
/check/python3.11-chainguard: ELF 64-bit LSB pie executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, for GNU/Linux 4.9.0, stripped
	linux-vdso.so.1 (0x0000ffffa5703000)
	libpython3.11.so.1.0 => not found
	libc.so.6 => /lib/aarch64-linux-gnu/libc.so.6 (0x0000ffffa54f0000)
	/lib/ld-linux-aarch64.so.1 (0x0000ffffa56ca000)
```

#### for Alpine Linux 3.17
```
Step 9/10 : RUN checksec --file=/check/python3.10-alpine &&  file /check/python3.10-alpine &&  ldd /check/python3.10-alpine
 ---> Running in 06182b5238fd
RELRO           STACK CANARY      NX            PIE             RPATH      RUNPATH	Symbols		FORTIFY	Fortified	Fortifiable	FILE
Full RELRO      No canary found   NX enabled    PIE enabled     No RPATH   No RUNPATH   No Symbols	  No	0		0		/check/python3.10-alpine
/check/python3.10-alpine: ELF 64-bit LSB pie executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-musl-aarch64.so.1, stripped
	linux-vdso.so.1 (0x0000ffff8e321000)
	libpython3.10.so.1.0 => not found
	libc.musl-aarch64.so.1 => not found
```

#### for Ubuntu Jammy
```
Step 10/10 : RUN checksec --file=/usr/bin/python3.10 &&  file /usr/bin/python3.10 &&  ldd /usr/bin/python3.10
 ---> Running in 6bdbd9cb571e
RELRO           STACK CANARY      NX            PIE             RPATH      RUNPATH	Symbols		FORTIFY	Fortified	Fortifiable	FILE
Full RELRO      Canary found      NX enabled    PIE enabled     No RPATH   No RUNPATH   No Symbols	  Yes	18		43		/usr/bin/python3.10
/usr/bin/python3.10: ELF 64-bit LSB pie executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, BuildID[sha1]=a1841de4f4ec9445a10bff638afa4c72deace9e0, for GNU/Linux 3.7.0, stripped
	linux-vdso.so.1 (0x0000ffffaca8d000)
	libm.so.6 => /lib/aarch64-linux-gnu/libm.so.6 (0x0000ffffac3f0000)
	libexpat.so.1 => /lib/aarch64-linux-gnu/libexpat.so.1 (0x0000ffffac3b0000)
	libz.so.1 => /lib/aarch64-linux-gnu/libz.so.1 (0x0000ffffac380000)
	libc.so.6 => /lib/aarch64-linux-gnu/libc.so.6 (0x0000ffffac1d0000)
	/lib/ld-linux-aarch64.so.1 (0x0000ffffaca54000)
```

### Expected output

Aligned with the results from Ubuntu >=Jammy.

**Comment by kaniini** [comment]
It seems like this is due to a bug in the Python build system when PGO is used.  Ubuntu has an interesting workaround for it.

**Comment by kaniini** [comment]
After further investigation, it appears that this is a false positive, and the reason why Ubuntu's `python3.10` binary passes your test is because they are explicitly building Python with the old `-fstack-protector` flag rather than modern `-fstack-protector-strong`, which omits stack canaries for stack frames which do not have buffers requiring protection.

Investigating this issue was interesting.  It turns out that the `python3.X` program is just a stub which jumps into `Py_BytesMain` as quickly as possible.  It does not even have a `main` function of its own: as you can see below, it jumps straight into `Py_BytesMain` when started.

<img width="739" alt="Screenshot 2023-02-09 at 6 10 47 PM" src="https://user-images.githubusercontent.com/1522444/217982684-7234334e-7e6b-434e-8195-637177613d35.png">

Because of this, there are no buffers on the stack requiring protection, and as there are no buffers, GCC does not generate any stack canaries, which is why checksec reports there are no canaries.  But there aren't supposed to be any.

Where does this `Py_BytesMain` come from, anyway?  Well, based on the above screenshot, we can see it is in the PLT, so it is provided by a `DT_NEEDED` library, e.g. `libpython3.11`.  If we run checksec on that:

<img width="1283" alt="Screenshot 2023-02-09 at 6 17 06 PM" src="https://user-images.githubusercontent.com/1522444/217983484-1d789035-be07-4ce0-89d7-1bced21782ed.png">

We see that FORTIFY and stack canaries are indeed present and in use.

The RELRO issue was solved in #557, and was caused by some packages not having `-z now` LDFLAGS explicitly in the environment.  As packages are rebuilt due to updates, this particular issue will resolve itself.  I do not believe it warrants a distribution-wide rebuild to ensure `-z now` is consistent.

**Comment by nrvnrvn** [comment]
Thanks a lot @kaniini for the deep dive and for adding support for Full RELRO.

Indeed, `-z now` is not enabled by default by linkers. So it must be explicitly passed when building packages or enabled by default when building linkers themselves.

As for stack canaries, checksec decides whether they are "found" or not by [analyzing the symbol table](https://github.com/slimm609/checksec.sh/blob/eae2732fe03695752ae234da634db005af12fa05/checksec#L708) which is apparently not the most accurate indicator. Because when there are no vulnerable functions to protect only `-fstack-protector-all` will _please_ checksec and protecting all functions comes with a huge performance penalty.
